kernel/aster-nix/src/syscall/dup.rs

```diff
@@ -1,12 +1,38 @@
 // SPDX-License-Identifier: MPL-2.0
 
-use super::{SyscallReturn, SYS_DUP, SYS_DUP2};
+use super::{SyscallReturn, SYS_DUP, SYS_DUP2, SYS_DUP3};
 use crate::{
-    fs::file_table::{FdFlags, FileDesc},
+    fs::file_table::{FdFlags, FileDesc, FileTable},
     log_syscall_entry,
     prelude::*,
+    process::ResourceType,
 };
 
+fn do_dup3(old_fd: FileDesc, new_fd: FileDesc, flags: FdFlags) -> Result<SyscallReturn> {
+    if old_fd == new_fd {
+        return_errno!(Errno::EINVAL);
+    }
+
+    let current = current!();
+    if new_fd
+        >= current
+            .resource_limits()
+            .lock()
+            .get_rlimit(ResourceType::RLIMIT_NOFILE)
+            .get_cur() as FileDesc
+    {
+        return_errno!(Errno::EBADF);
+    }
+
+    let mut file_table: MutexGuard<FileTable> = current.file_table().lock();
+    let file = file_table.get_file(old_fd)?.clone();
+    if let Some(old_file) = file_table.insert_at(new_fd, file, flags) {
+        let _ = old_file.clean_for_close();
```

Comment:
Drop the lock on file table before calling `clean_for_close`. Because there is a possibility that `clean_for_close` also wants to operate on the file table.

Reply:
I will make `old_file` as the return value of the code block.